src/constraint.rs

```diff
@@ -1274,3 +1284,66 @@ impl ConstraintValidator for AnnotationsConstraint {
         self.validate_unordered_annotations(value, type_store, violations)
     }
 }
+
+/// Implements an `precision` constraint of Ion Schema
+/// [precision]: https://amzn.github.io/ion-schema/docs/spec.html#precision
+#[derive(Debug, Clone, PartialEq)]
+pub struct PrecisionConstraint {
+    precision_range: Range,
+}
+
+impl PrecisionConstraint {
+    pub fn new(length_range: Range) -> Self {
+        Self {
+            precision_range: length_range,
+        }
+    }
+
+    pub fn precision(&self) -> &Range {
+        &self.precision_range
+    }
+}
+
+impl ConstraintValidator for PrecisionConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // get the precision of given decimal
+        let value_precision = match value.as_decimal() {
+            Some(decimal_value) => decimal_value.precision(),
+            _ => {
+                // return Violation if value is not decimal
+                return Err(Violation::new(
+                    "precision",
+                    ViolationCode::TypeMismatched,
+                    &format!(
+                        "expected a decimal but found {}",
+                        if value.is_null() {
+                            format!("{:?}", value)
+                        } else {
+                            format!("{}", value.ion_type())
+                        }
+                    ),
```

Comment:
This works, but it requires two String allocations instead of one. I suggest:

```rust
if value.is_null() {
   format!("expected a decimal but found {:?}", value);
} else {
   format!("expected a decimal but found {}", value.ion_type());
}
```

Even better: `OwnedElement` should implement `Display` so we could just print the value regardless of whether it's null.